projects/aws/lib/step-function.ts

```diff
@@ -0,0 +1,113 @@
+import cdk = require('@aws-cdk/core')
+import lambda = require('@aws-cdk/aws-lambda')
+import { Code, FunctionProps } from '@aws-cdk/aws-lambda'
+import s3 = require('@aws-cdk/aws-s3')
+import iam = require('@aws-cdk/aws-iam')
+import sfn = require('@aws-cdk/aws-stepfunctions')
+import tasks = require('@aws-cdk/aws-stepfunctions-tasks')
+import { toTitleCase } from './tools'
+import { Duration } from '@aws-cdk/core'
+interface StepFunctionProps {
+    stack: string
+    stage: string
+    deployBucket: s3.IBucket
+    publishedBucket: s3.IBucket
+    backendURL: string
+}
+
+const taskλ = (
+    name: string,
+    {
+        scope,
+        stack,
+        stage,
+        deployBucket,
+        publishedBucket,
+    }: {
+        scope: cdk.Construct
+        stack: string
+        stage: string
+        deployBucket: s3.IBucket
+        publishedBucket: s3.IBucket
+    },
+    environment?: { [key: string]: string },
+    overrides?: Partial<FunctionProps>,
+) => {
+    return new lambda.Function(scope, `EditionsArchiver${toTitleCase(name)}`, {
+        ...{
+            functionName: `editions-archiver-stepmachine-${name}-${stage}`,
+            runtime: lambda.Runtime.NODEJS_10_X,
+            timeout: Duration.minutes(5),
+            memorySize: 1500,
+            code: Code.bucket(
+                deployBucket,
+                `${stack}/${stage}/archiver/archiver.zip`,
+            ),
+            handler: `index.${name}`,
+            environment: {
+                stage: stage,
+                bucket: publishedBucket.bucketName,
+                ...environment,
+            },
+            initialPolicy: [
+                new iam.PolicyStatement({
+                    actions: ['*'],
+                    resources: [
+                        publishedBucket.arnForObjects('*'),
+                        publishedBucket.bucketArn,
+                    ],
+                }),
+            ],
+        },
+        ...overrides,
+    })
+}
+
+export const stepFunction = (
+    scope: cdk.Construct,
+    {
+        stack,
+        stage,
+        deployBucket,
+        publishedBucket,
+        backendURL,
+    }: StepFunctionProps,
+) => {
+    const issue = taskλ(
+        'issue',
+        {
+            scope,
+            stack,
+            stage,
+            deployBucket,
+            publishedBucket,
+        },
+        { backend: backendURL },
+    )
+
+    //Archiver step function
+    const issueTask = new sfn.Task(scope, 'Archive Job', {
+        task: new tasks.InvokeFunction(issue),
+    })
+
+    const hasFinished = new sfn.Choice(scope, 'Has finished', {})
+
+    hasFinished.when(
+        sfn.Condition.booleanEquals('$.finished', false),
+        issueTask,
+    )
+    hasFinished.otherwise(new sfn.Succeed(scope, 'successfully-archived'))
+
+    issueTask.next(hasFinished)
```

Comment:
Is this where it's calling itself recursively? I'm not to au fait with this stuff, but just to make me clearer!

Reply:
Yeah, going to move this into the calling lambda in another PR so it's all cleaner